File: src/core/services/commission_calculator.py
from decimal import Decimal, ROUND_HALF_UP
from src.core.domain.value_objects import Money, CommissionCalculationResult


class DynamicCommissionCalculator:
    """
    Calculador de comisiones dinámicas basado en el umbral monetario:
    - Monto < 300 BOB: 15% de comisión
    - Monto >= 300 BOB: 10% de comisión
    """

    THRESHOLD_BOB = Decimal("300.00")
    RATE_MICROPAYMENT = Decimal("0.15")
    RATE_STANDARD = Decimal("0.10")

    def calculate(self, gross_amount: Money) -> CommissionCalculationResult:
        if gross_amount.amount <= Decimal("0.00"):
            raise ValueError("El monto a calcular debe ser estrictamente mayor a 0.")

        if gross_amount.amount < self.THRESHOLD_BOB:
            rate = self.RATE_MICROPAYMENT
        else:
            rate = self.RATE_STANDARD

        commission_val = (gross_amount.amount * rate).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
        net_val = gross_amount.amount - commission_val

        return CommissionCalculationResult(
            gross_amount=gross_amount,
            commission_rate=rate,
            commission_amount=Money(commission_val, gross_amount.currency),
            net_consultant_amount=Money(net_val, gross_amount.currency)
        )

File: backend/src/domain/commissionCalculator.js
/**
 * Calculador de comisiones dinámicas de plataforma01
 * - Monto < 300 BOB: 15% de comisión
 * - Monto >= 300 BOB: 10% de comisión
 */
export class CommissionCalculator {
  static THRESHOLD_BOB = 300.0;
  static RATE_MICROPAYMENT = 0.15;
  static RATE_STANDARD = 0.10;

  static calculate(grossAmount) {
    const amount = Number(grossAmount);
    if (isNaN(amount) || amount <= 0) {
      throw new Error("El monto a calcular debe ser un número estrictamente mayor a 0.");
    }

    const rate = amount < this.THRESHOLD_BOB ? this.RATE_MICROPAYMENT : this.RATE_STANDARD;
    const commissionAmount = Math.round(amount * rate * 100) / 100;
    const netAmount = Math.round((amount - commissionAmount) * 100) / 100;

    return {
      grossAmount: amount,
      currency: "BOB",
      rate,
      ratePercentage: `${rate * 100}%`,
      commissionAmount,
      netAmount
    };
  }
}
